tracer/src/Datadog.Trace/OTelMetrics/MetricReaderHandler.cs

```diff
@@ -0,0 +1,340 @@
+                    Log.Debug("Enabled measurement events for instrument: {InstrumentName} from meter: {MeterName}", instrument.Name, meterName);
+                }
+                else
+                {
+                    Log.Warning("Failed to create MetricState for instrument '{InstrumentName}' from meter '{MeterName}'. Instrument will be ignored.", instrument.Name, meterName);
+                }
+            }
+        }
+
+        public static void OnMeasurementRecordedLong(Instrument instrument, long value, ReadOnlySpan<KeyValuePair<string, object?>> tags, object? state)
+        {
+            if (state is MetricState metricState)
+            {
+                metricState.RecordMeasurementLong(value, tags);
+            }
+            else
+            {
+                // Log error - this should not happen if we follow OTel SDK pattern
+                Log.Warning("Measurement recorded for instrument '{InstrumentName}' but no MetricState provided. Measurement will be dropped.", instrument.Name);
+            }
+        }
+
+        public static void OnMeasurementRecordedDouble(Instrument instrument, double value, ReadOnlySpan<KeyValuePair<string, object?>> tags, object? state)
+        {
+            if (state is MetricState metricState)
+            {
+                metricState.RecordMeasurementDouble(value, tags);
+            }
+            else
+            {
+                // Log error - this should not happen if we follow OTel SDK pattern
+                Log.Warning("Measurement recorded for instrument '{InstrumentName}' but no MetricState provided. Measurement will be dropped.", instrument.Name);
+            }
+        }
+
+        // Overloads for different numeric types
+        public static void OnMeasurementRecordedByte(Instrument instrument, byte value, ReadOnlySpan<KeyValuePair<string, object?>> tags, object? state)
+        {
+            OnMeasurementRecordedLong(instrument, value, tags, state);
+        }
+
+        public static void OnMeasurementRecordedShort(Instrument instrument, short value, ReadOnlySpan<KeyValuePair<string, object?>> tags, object? state)
+        {
+            OnMeasurementRecordedLong(instrument, value, tags, state);
+        }
+
+        public static void OnMeasurementRecordedInt(Instrument instrument, int value, ReadOnlySpan<KeyValuePair<string, object?>> tags, object? state)
+        {
+            OnMeasurementRecordedLong(instrument, value, tags, state);
+        }
+
+        public static void OnMeasurementRecordedFloat(Instrument instrument, float value, ReadOnlySpan<KeyValuePair<string, object?>> tags, object? state)
+        {
+            OnMeasurementRecordedDouble(instrument, value, tags, state);
+        }
+
+        internal static IReadOnlyList<MetricPoint> GetMetricsForExport(bool clearAfterGet = false)
+        {
+            var allMetricPoints = new List<MetricPoint>();
+
+            foreach (var kvp in CapturedMetrics)
```

Comment:
According to the docs:

> The enumerator returned from the dictionary is safe to use concurrently with reads and writes to the dictionary, however it does not represent a moment-in-time snapshot of the dictionary. The contents exposed through the enumerator may contain modifications made to the dictionary after [GetEnumerator](https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.getenumerator?view=net-9.0) was called.

I wonder if we should be working with a snapshot, to avoid missing metrics. They could be being added while you're enumerating, get missed, and then get cleared 🤔